File: engine/ledger.py
"""Cursor tracking, claim/finalize dedup, failure-issue tracking, and heartbeat.

All business logic lives here, backend-agnostic — it only ever calls
`StateBackend.load`/`save`. This is what makes it possible to point the exact
same logic at local JSON files (GitHub Actions edition) or, later, at SQLite
(self-hosted edition) with zero changes to this file.
"""
from __future__ import annotations

from datetime import datetime, timedelta
from typing import Optional

from .state_backend import StateBackend
from .timeutil import iso, parse_iso

DEFAULT_LOOKBACK = timedelta(minutes=15)
DEDUP_RETENTION = timedelta(days=7)
STALE_CLAIM_GRACE = timedelta(minutes=10)


class Ledger:
    def __init__(self, backend: StateBackend):
        self.backend = backend
        self.cursors: dict = {}
        self.dedup: dict = {}
        self.issues: dict = {}
        self.heartbeat: dict = {}
        self._dirty: set[str] = set()

    def load(self) -> None:
        self.cursors = self.backend.load("state/cursors", {})
        self.dedup = self.backend.load("state/dedup", {"executed": {}})
        self.issues = self.backend.load("state/issues", {})
        self.heartbeat = self.backend.load("state/heartbeat", {})
        self._dirty.clear()

    # ---- cursors ----

    def get_cursor(self, job_id: str, now: datetime, lookback: timedelta = DEFAULT_LOOKBACK) -> datetime:
        """Persisted watermark if we've evaluated this job before; otherwise a
        short rolling lookback (never backfills from job-creation time)."""
        entry = self.cursors.get(job_id)
        if entry is None:
            return now - lookback
        return parse_iso(entry["last_evaluated_utc"])

    def advance_cursor(self, job_id: str, when: datetime, status: str) -> None:
        prev = self.cursors.get(job_id, {})
        consecutive_failures = prev.get("consecutive_failures", 0)
        if status == "failed":
            consecutive_failures += 1
        elif status == "success":
            consecutive_failures = 0
        self.cursors[job_id] = {
            "last_evaluated_utc": iso(when),
            "last_status": status,
            "consecutive_failures": consecutive_failures,
        }
        self._dirty.add("state/cursors")

    def consecutive_failures(self, job_id: str) -> int:
        """Failures in a row as of the *last* advance_cursor call — call this
        BEFORE advancing the cursor with a new result if you need to detect a
        transition into/out of a failing streak (e.g. for recovery notifications)."""
        return self.cursors.get(job_id, {}).get("consecutive_failures", 0)

    # ---- dedup / claim-finalize ----

    @staticmethod
    def _key(job_id: str, scheduled_time: datetime) -> str:
        return f"{job_id}|{iso(scheduled_time)}"

    def is_claimed(self, job_id: str, scheduled_time: datetime) -> bool:
        """True if this occurrence has ever been claimed, regardless of outcome.
        We never re-execute a claimed occurrence — we can't know whether an
        interrupted run's side effect actually fired, so "maybe already ran" is
        treated the same as "definitely already ran". `reconcile_stale_claims`
        is what cleans up abandoned "running" entries, not re-execution."""
        return self._key(job_id, scheduled_time) in self.dedup["executed"]

    def claim(self, job_id: str, scheduled_time: datetime, now: datetime) -> None:
        self.dedup["executed"][self._key(job_id, scheduled_time)] = {
            "status": "running",
            "claimed_at": iso(now),
            "finished_at": None,
        }
        self._dirty.add("state/dedup")

    def reconcile_stale_claims(self, now: datetime, grace: timedelta = STALE_CLAIM_GRACE) -> list[str]:
        """Flip claims stuck in 'running' past `grace` (dispatcher was killed
        mid-execution) to 'failed' so they stop looking perpetually in-flight.
        Returns the affected dedup keys for logging/heartbeat visibility."""
        affected = []
        for key, entry in self.dedup["executed"].items():
            if entry["status"] == "running":
                claimed_at = parse_iso(entry["claimed_at"])
                if (now - claimed_at) >= grace:
                    entry["status"] = "failed"
                    entry["finished_at"] = iso(now)
                    affected.append(key)
        if affected:
            self._dirty.add("state/dedup")
        return affected

    def finalize(self, job_id: str, scheduled_time: datetime, status: str, finished_at: datetime) -> None:
        key = self._key(job_id, scheduled_time)
        entry = self.dedup["executed"].setdefault(
            key, {"status": "running", "claimed_at": iso(finished_at)}
        )
        entry["status"] = status
        entry["finished_at"] = iso(finished_at)
        self._dirty.add("state/dedup")

    def prune_dedup(self, now: datetime) -> None:
        cutoff = now - DEDUP_RETENTION
        executed = self.dedup["executed"]
        kept = {}
        for key, entry in executed.items():
            ts_str = entry.get("finished_at") or entry.get("claimed_at")
            ts = parse_iso(ts_str) if ts_str else now
            if ts >= cutoff:
                kept[key] = entry
        if len(kept) != len(executed):
            self._dirty.add("state/dedup")
        self.dedup["executed"] = kept

    # ---- failure issue tracking ----

    def open_issue_number(self, job_id: str) -> Optional[int]:
        return self.issues.get(job_id, {}).get("issue_number")

    def record_open_issue(self, job_id: str, issue_number: int, opened_at: datetime) -> None:
        self.issues[job_id] = {"issue_number": issue_number, "opened_at": iso(opened_at)}
        self._dirty.add("state/issues")

    def clear_issue(self, job_id: str) -> None:
        if job_id in self.issues:
            del self.issues[job_id]
            self._dirty.add("state/issues")

    # ---- heartbeat ----

    def record_heartbeat(self, now: datetime, status: str, detail: str = "") -> None:
        self.heartbeat = {"last_run_utc": iso(now), "status": status, "detail": detail[:200]}
        self._dirty.add("state/heartbeat")

    # ---- persistence ----

    def dirty_names(self) -> set[str]:
        return set(self._dirty)

    def flush(self, only: Optional[set[str]] = None) -> list[str]:
        """Persist dirty blobs (optionally restricted to `only`), returning the
        names actually written so the caller (a git-committing adapter) knows
        what changed."""
        names = self._dirty if only is None else (self._dirty & only)
        written = []
        for name in sorted(names):
            data = {
                "state/cursors": self.cursors,
                "state/dedup": self.dedup,
                "state/issues": self.issues,
                "state/heartbeat": self.heartbeat,
            }[name]
            self.backend.save(name, data)
            written.append(name)
        self._dirty -= names
        return written

File: src/engine/ledger.js
"use strict";

const { assertStateBackend } = require("./state-backend");
const { iso, parseIso } = require("./timeutil");

const DEFAULT_LOOKBACK_MS = 15 * 60 * 1000;
const DEDUP_RETENTION_MS = 7 * 24 * 60 * 60 * 1000;
const STALE_CLAIM_GRACE_MS = 10 * 60 * 1000;

class Ledger {
  constructor(backend) {
    this.backend = assertStateBackend(backend, "Ledger backend");
    this.cursors = {};
    this.dedup = {};
    this.issues = {};
    this.heartbeat = {};
    this.dirty = new Set();
  }

  load() {
    this.cursors = this.backend.load("state/cursors", {});
    this.dedup = this.backend.load("state/dedup", { executed: {} });
    this.issues = this.backend.load("state/issues", {});
    this.heartbeat = this.backend.load("state/heartbeat", {});
    this.dirty.clear();
  }

  getCursor(jobId, now, lookbackMs = DEFAULT_LOOKBACK_MS) {
    const entry = this.cursors[jobId];
    if (!entry) return new Date(now.getTime() - lookbackMs);
    return parseIso(entry.last_evaluated_utc);
  }

  healthPatch(status, extra = {}) {
    const patch = { last_status: status };
    if (status === "success") {
      patch.consecutive_failures = 0;
      patch.failure_pause_until_utc = null;
      patch.auto_disabled = false;
      patch.auto_disabled_at_utc = null;
      patch.auto_disabled_reason = null;
    } else if (extra.consecutive_failures !== undefined) {
      patch.consecutive_failures = extra.consecutive_failures;
    }
    return { ...patch, ...extra };
  }

  recordJobHealth(jobId, when, status, extra = {}) {
    const prev = this.cursors[jobId] || {};
    let consecutiveFailures = prev.consecutive_failures || 0;
    if (status === "failed") consecutiveFailures += 1;
    if (status === "success") consecutiveFailures = 0;
    this.cursors[jobId] = { ...prev, ...this.healthPatch(status, { consecutive_failures: consecutiveFailures, ...extra }) };
    this.dirty.add("state/cursors");
  }

  advanceCursor(jobId, when, status, extra = {}) {
    this.recordJobHealth(jobId, when, status, { ...extra, last_evaluated_utc: iso(when) });
  }

  consecutiveFailures(jobId) {
    return (this.cursors[jobId] || {}).consecutive_failures || 0;
  }

  isAutoDisabled(jobId) {
    return Boolean((this.cursors[jobId] || {}).auto_disabled);
  }

  failurePauseUntil(jobId) {
    const value = (this.cursors[jobId] || {}).failure_pause_until_utc;
    return value ? parseIso(value) : null;
  }

  key(jobId, scheduledTime) {
    return `${jobId}|${iso(scheduledTime)}`;
  }

  isClaimed(jobId, scheduledTime) {
    return Boolean(this.dedup.executed[this.key(jobId, scheduledTime)]);
  }

  claim(jobId, scheduledTime, now) {
    this.dedup.executed[this.key(jobId, scheduledTime)] = {
      status: "running",
      claimed_at: iso(now),
      finished_at: null,
    };
    this.dirty.add("state/dedup");
  }

  reconcileStaleClaims(now, graceMs = STALE_CLAIM_GRACE_MS) {
    const affected = [];
    for (const [key, entry] of Object.entries(this.dedup.executed)) {
      if (entry.status !== "running") continue;
      const claimedAt = parseIso(entry.claimed_at);
      if (now.getTime() - claimedAt.getTime() >= graceMs) {
        entry.status = "failed";
        entry.finished_at = iso(now);
        affected.push(key);
      }
    }
    if (affected.length) this.dirty.add("state/dedup");
    return affected;
  }

  finalize(jobId, scheduledTime, status, finishedAt) {
    const key = this.key(jobId, scheduledTime);
    const entry = this.dedup.executed[key] || { status: "running", claimed_at: iso(finishedAt) };
    entry.status = status;
    entry.finished_at = iso(finishedAt);
    this.dedup.executed[key] = entry;
    this.dirty.add("state/dedup");
  }

  pruneDedup(now) {
    const cutoff = now.getTime() - DEDUP_RETENTION_MS;
    const kept = {};
    for (const [key, entry] of Object.entries(this.dedup.executed)) {
      const stamp = entry.finished_at || entry.claimed_at;
      const ts = stamp ? parseIso(stamp).getTime() : now.getTime();
      if (ts >= cutoff) kept[key] = entry;
    }
    if (Object.keys(kept).length !== Object.keys(this.dedup.executed).length) {
      this.dirty.add("state/dedup");
    }
    this.dedup.executed = kept;
  }

  openIssueNumber(jobId) {
    return (this.issues[jobId] || {}).issue_number ?? null;
  }

  recordOpenIssue(jobId, issueNumber, openedAt) {
    this.issues[jobId] = { issue_number: issueNumber, opened_at: iso(openedAt) };
    this.dirty.add("state/issues");
  }

  clearIssue(jobId) {
    if (this.issues[jobId]) {
      delete this.issues[jobId];
      this.dirty.add("state/issues");
    }
  }

  recordHeartbeat(now, status, detail = "") {
    this.heartbeat = { last_run_utc: iso(now), status, detail: detail.slice(0, 200) };
    this.dirty.add("state/heartbeat");
  }

  dirtyNames() {
    return new Set(this.dirty);
  }

  flush(only = null) {
    const names = [...this.dirty].filter((name) => !only || only.has(name)).sort();
    const dataByName = {
      "state/cursors": this.cursors,
      "state/dedup": this.dedup,
      "state/issues": this.issues,
      "state/heartbeat": this.heartbeat,
    };
    for (const name of names) this.backend.save(name, dataByName[name]);
    for (const name of names) this.dirty.delete(name);
    return names;
  }
}

module.exports = { Ledger, DEFAULT_LOOKBACK_MS, DEDUP_RETENTION_MS, STALE_CLAIM_GRACE_MS };
